src/rpcq.lisp

```diff
@@ -533,27 +208,61 @@ We distinguish between the following options for any field type:
     tbl))
 
 
-(defmacro defmessage (class-name field-specs &key (documentation nil))
-  "Create a (de-)serializable message definition with name CLASS-NAME and slots (SLOT-SPECS)."
+(defmacro defmessage (class-name parent-name field-specs &key (documentation nil))
+  "Create a (de-)serializable message definition.  Its class name is CLASS-NAME; its parent message class is named by the list PARENT-NAMES (which is neither () or of length 1); and its slots are specified by FIELD-SPECS.
+
+A an entry in FIELD-SPECS is a plist with the following possible keys:
+
+* :TYPE          - Either an atomic type, a list type, a map type, or another RPCQ message name.
+* :REQUIRED      - A boolean indicating whether the field is required. If it is required, there must be a default value, either supplied through :DEFAULT or implicitly.
+* :DOCUMENTATION - A documentation string.
+* :DEFAULT       - A default value for the slot.
+* :DEPRECATES    - The slot name that this slot replaces.
+* :DEPRECATED-BY - The slot name that this slot is replaced by.
+* :DEPRECATED    - Indicates that this slot is being phased out without a replacement.
+
+LIMITATIONS:
+  * DEPRECATES / DEPRECATED-BY must come in balanced pairs.
+  * DEPRECATES / DEPRECATED-BY are mutually exclusive. (Nested deprecation is not currently supported.)
+  * DEPRECATES / DEPRECATED-BY must reference slots on the same message, and not on parent messages.
+"
   (check-type class-name symbol)
-  (setf *messages* (nconc *messages* `((,class-name ,field-specs ,documentation))))
-  (flet ((accessor (slot-name)
-           (alexandria:symbolicate (symbol-name class-name)
-                                   "-"
-                                   (symbol-name slot-name))))
-    (flet ((make-slot-spec (field-spec)
+  (assert (or (null parent-name)
+              (and (typep parent-name 'cons)
+                   (= 1 (length parent-name)))))
+  (setf *messages* (nconc *messages* `((,class-name ,(first parent-name) ,field-specs ,documentation))))
```

Comment:
Why not `(push (list class-name (first ..) ..) *messages*)`? Is ordering important? Maybe `*messages*` should be a table anyway.

Reply:
Ordering is important at the final stage: Lisp is fine having message definitions reference each other (as types) willy-nilly, but python expects messages with typed fields to have those types declared before they're referenced. Either we have to track that data ourselves, or we expect the user to order their spec for us and we track their ordering.